core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -70,45 +72,28 @@ public LockCache(Function<? super K, ? extends ReentrantReadWriteLock> defaultLo
     mHardLimit = maxSize;
     mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
     mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
-    mIterator = mCache.entrySet().iterator();
-    mEvictLock = new ReentrantLock();
-    mLastSizeWarningTime = 0;
   }
 
   /**
-   * If the size of the cache exceeds the soft limit and no other thread is evicting entries,
-   * start evicting entries.
+   * If the size of the cache exceeds the soft limit, evict entries with zero references.
    */
   private void evictIfOverLimit() {
-    int numToEvict = mCache.size() - mSoftLimit;
-
-    if (numToEvict <= 0) {
+    if (mCache.size() <= mSoftLimit) {
       return;
     }
-
-    if (mEvictLock.tryLock()) {
-      try {
-        // update the total number to evict while we are holding the lock
-        numToEvict = mCache.size() - mSoftLimit;
-        // This thread won the race as the evictor
-        while (numToEvict > 0) {
-          if (!mIterator.hasNext()) {
-            mIterator = mCache.entrySet().iterator();
-          }
-          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
-          ValNode candidate = candidateMapEntry.getValue();
-          if (candidate.mIsAccessed) {
-            candidate.mIsAccessed = false;
-          } else {
-            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
-              // the value object can be evicted, at the same time we make refCount minValue
-              mIterator.remove();
-              numToEvict--;
-            }
-          }
+    try (LockResource l = new LockResource(mEvictLock)) {
```

Comment:
should this be a try-lock, since blocking will stop all progress for other threads?

Reply:
I intentionally make it block other threads since I don't want to lose chances for eviction.
But after second thought, try-lock works too.